registry/storage/driver/s3-aws/s3.go

```diff
@@ -1116,12 +1117,28 @@ func (d *driver) doWalk(parentCtx context.Context, objectCount *int64, path, pre
 		walkInfos := make([]storagedriver.FileInfoInternal, 0, len(objects.Contents))
 
 		for _, file := range objects.Contents {
+			filePath := strings.Replace(*file.Key, d.s3Path(""), prefix, 1)
+
+			// get a list of all inferred directories skipped between the previous directory and this file
+			dirs := directoryDiff(prevDir, filePath)
+			if len(dirs) > 0 {
+				for _, dir := range dirs {
+					walkInfos = append(walkInfos, storagedriver.FileInfoInternal{
+						FileInfoFields: storagedriver.FileInfoFields{
+							IsDir: true,
+							Path:  dir,
+						},
+					})
+					prevDir = dir
```

Comment:
If there is more than one element in `dirs`, would we end up with duplicate `walkInfos` elements for some of those directories since only the last one gets assigned to `prevDir` and passed in to `directoryDiff` on the next object? So, supposing we had two deeply nested files:

```
/meow/hello/world.txt
/meow/hello/world/foo/bar/baz/bonk.txt
/meow/hello/world/foo/bar/baz/boink.txt
```

If I understand correctly with this would result in the following `walkInfos`:

* `/meow/hello`
* `/meow/hello/world/`
* `/meow/hello/world/foo`
* `/meow/hello/world/foo/bar`
* `/meow/hello/world/foo/bar/baz`
* `/meow/hello/world/foo/bar/baz/bonk`
* `/meow/hello/world/foo`
* `/meow/hello/world/foo/bar`
* `/meow/hello/world/foo/bar/baz`
* `/meow/hello/world/foo/bar/baz/boink`

Reply:
`prevDir = dir` is done intentionally as a simple way to save the last dir, which ends up being the parent of the object (or is not re-assigned if directoryDiff yields empty results). This doesn't cause any duplicates though, the unit tests should cover a good amount of cases. 

```
directoryDiff("/meow/hello", "/meow/hello/world/foo/bar/baz/bonk.txt") 
   => ["/meow/hello/world", "/meow/hello/world/foo", "/meow/hello/world/foo/bar", "/meow/hello/world/foo/bar/baz"]
```
results in `prevDir = "/meow/hello/world/foo/bar/baz"`

Then next iteration 
```
directoryDiff("/meow/hello/world/foo/bar/baz", "/meow/hello/world/foo/bar/baz/boink.txt") 
   => [ ]
``` 
because they both belong to the same directory, no new directory `walkInfo` are added. `prevDir` is left as ` "/meow/hello/world/foo/bar/baz"`

Reply:
Ah, I was misunderstanding the break logic in `directoryDiff` - I see how it works now.